From 76644bc551f583c905277b0a826b6c5efb3ba04c Mon Sep 17 00:00:00 2001
From: Zheng Zucheng <zhengzucheng@huawei.com>
Date: Wed, 11 May 2022 09:23:28 +0800
Subject: [PATCH 2070/2592] [Backport] cputime: fix undefined reference to
 get_idle_time when CONFIG_PROC_FS disabled

hulk inclusion
category: bugfix
bugzilla: 13257

Reference: https://gitee.com/openeuler/kernel/commit/1644ea3f28fdabe06b76afda2a35e2212dafc673

-------------------------------

Add get_idle_time stub function when CONFIG_PROC_FS disabled,
to avoid this compile error.

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 include/linux/kernel_stat.h | 8 ++++++++
 kernel/sched/cputime.c      | 5 +++++
 2 files changed, 13 insertions(+)

diff --git a/include/linux/kernel_stat.h b/include/linux/kernel_stat.h
index d089ef530e04..471b20fac4a6 100644
--- a/include/linux/kernel_stat.h
+++ b/include/linux/kernel_stat.h
@@ -105,7 +105,15 @@ extern void account_system_index_time(struct task_struct *, u64,
 				      enum cpu_usage_stat);
 extern void account_steal_time(u64);
 extern void account_idle_time(u64);
+
+#ifdef CONFIG_PROC_FS
 extern u64 get_idle_time(struct kernel_cpustat *kcs, int cpu);
+#else
+static inline u64 get_idle_time(struct kernel_cpustat *kcs, int cpu)
+{
+	return -1ULL;
+}
+#endif
 
 #ifdef CONFIG_VIRT_CPU_ACCOUNTING_NATIVE
 static inline void account_process_tick(struct task_struct *tsk, int user)
diff --git a/kernel/sched/cputime.c b/kernel/sched/cputime.c
index c31ca8d5c0b0..44a6e41bc3aa 100644
--- a/kernel/sched/cputime.c
+++ b/kernel/sched/cputime.c
@@ -1106,6 +1106,11 @@ int sched_idle_time_adjust(int cpu, u64 *utime, u64 *stime)
 
 	raw_spin_lock(&rq_cputime->lock);
 	kcpustat_cpu_fetch(&kcs, cpu);
+	/* If failed to get idle time, drop adjustment */
+	if (get_idle_time(&kcs, cpu) == -1ULL) {
+		raw_spin_unlock(&rq_cputime->lock);
+		return 0;
+	}
 
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
 	if (sched_clock_irqtime) {
-- 
2.17.1

